include/swift/SIL/ReborrowVerifier.h

```diff
@@ -0,0 +1,105 @@
+//===--- ReborrowVerifier.h ------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SIL_REBORROWVERIFIER_H
+#define SWIFT_SIL_REBORROWVERIFIER_H
+
+#include "swift/SIL/LinearLifetimeCheckerPrivate.h"
+#include "swift/SIL/SILInstruction.h"
+#include "swift/SIL/SILValue.h"
+
+namespace swift {
+
+class DeadEndBlocks;
+
+// A guaranteed phi arg ends the borrow scope of its incoming value and begins a
+// new borrow scope. ReborrowVerifier validates the lifetime of the reborrow
+// lies within the lifetime of its base value. It uses LinearLifetimeChecker for
+// this.
+class ReborrowVerifier {
+  /// worklist of reborrows and their base value
+  llvm::SmallVector<std::pair<SILPhiArgument *, SILValue>, 8> worklist;
+
+  /// A cache of dead-end basic blocks that we use to determine if we can
+  /// ignore "leaks".
+  DeadEndBlocks &deadEndBlocks;
+
+  /// A cache map of borrow lifetime ending operands and their base value
+  llvm::DenseMap<Operand *, SILValue> visitedOps;
+  /// A cache map of guaranteed phi args and their base value
+  llvm::DenseMap<SILPhiArgument *, SILValue> visitedPhiArgs;
+
+  llvm::SmallPtrSet<SILPhiArgument*, 4> phiArgsThatRequireNoLifetimeEndUses;
+
+  /// The builder that the checker uses to emit error messages, crash if asked
+  /// for, or supply back interesting info to the caller.
+  LinearLifetimeChecker::ErrorBuilder errorBuilder;
+
+public:
+  ReborrowVerifier(const SILFunction *func, DeadEndBlocks &deadEndBlocks,
+                   LinearLifetimeChecker::ErrorBuilder errorBuilder)
+      : deadEndBlocks(deadEndBlocks), errorBuilder(errorBuilder) {}
+
+  // Verifies all reborrows of a root value lie within the lifetime of its base
+  // values.
+  bool verify();
+
+  // Add a guaranteed phiArg and its base value to the worklist for verification
+  void add(SILPhiArgument *phiArg, SILValue baseVal) {
```

Comment:
Does this need to be in SIL? Can we hide it as a library detail?

Reply:
I'll update it.

Reply:
Done.